### PR TITLE
Introduce Bethe Free Energy

### LINKED ISSUE #152: Introduce Bethe Free Energy and variational inference concept in the documentation
The current documentation for RxInfer lacks clarity regarding the concept of Bethe Free Energy and does not sufficiently address the importance of working with it. Additionally, the documentation does not provide comprehensive information about the concept of variational inference. The existing [documentation page](https://biaslab.github.io/RxInfer.jl/stable/library/bethe-free-energy/) dedicated to Bethe Free Energy fails to provide a straightforward explanation of its essence.

We require an introductory section in the documentation that offers simple, plain-language explanations for the following concepts:

- Variational inference.
- Variational Free Energy.
- Bethe Free Energy constraints.
- 
This addition will enhance the documentation's accessibility and clarity for users.

### PR DESCRIPTION
**Pull request by ThijsvdLaar**
This branch includes an introduction to the Bethe Free Energy in the docs. Fixes #152 and #95.

### CODE REVIEW

File: docs/src/library/bethe-free-energy.md [hunk 1]
```diff
@@ -1,5 +1,71 @@
...
+Individual factors may represent stochastic functions, such as conditional or prior distributions, but also potential functions or deterministic relationships. A factor may depend on multiple observed and/or latent variables (or none).
+
+
+## [Variational Free Energy](@id lib-bethe-vfe)
+
+The Variational Free Energy (VFE) defines a functional objective that includes the model and a variational distribution over the latent variables.
+
+$$F[q](\hat{y}) = \mathbb{E}_{q(x)}\left[\log \frac{q(x)}{f(y=\hat{y}, x)} \right]$$
+
+A functional is the mathematical term for a function of a function that returns a scalar.
```
Comment (bvdmitri) [inline docs/src/library/bethe-free-energy.md:28]: Are these spaces between the sentences intentional? Markdown starts a new paragraph if there is a blank line (as Latex)

Comment (bvdmitri) [reply docs/src/library/bethe-free-energy.md:28]: Screenshot 2023-10-30 at 10 50 12

File: docs/src/library/bethe-free-energy.md [hunk 2]
```diff
@@ -1,5 +1,71 @@
...
+The mathematical technique used for this optimization is called variational calculus.
+
+The variational distribution is optimized under constraints that enforce e.g. the normalization of the variational distribution. The variational distribution that minimizes the VFE then approximates the true (but often unobtainable) posterior distribution.
+
+## [Bethe approximation](@id lib-bethe-approximation)
+Optimization of the VFE is still a daunting task, because the variational distribution is a joint distribution over possibly many latent variables. Instead of optimizing the joint variational distribution directly, a factorized variational distribution is often chosen, and optimized for its constituent factors.
+
+A popular choice of factorization is the Bethe approximation, which is constructed from the factorization of the model itself.
+
+$$q(x) \approx \frac{\prod_a q_a(x_a)}{\prod_i q_i(x_i)^{d_i - 1}}\,.$$
```
Comment (bvdmitri) [inline docs/src/library/bethe-free-energy.md:50]: Why `\approx`? Shouldn't it be just `=` as its our own choice

Comment (bvdmitri) [reply docs/src/library/bethe-free-energy.md:50]: What is `d_i`

File: docs/src/library/bethe-free-energy.md [hunk 3]
```diff
@@ -1,5 +1,71 @@
...
+
+## [Bethe Free Energy](@id lib-bethe-bfe)
+
+The Bethe Free Energy (BFE) substitutes the Bethe approximation in the VFE, which then fragments over factors and variables.
+
+$$F_B[q](\hat{y}) = \sum_a U_a[q_a](\hat{y}_a) - \sum_a H[q_a] + \sum_i (d_i - 1) H[q_i]$$
+
+The first term of the BFE specifies an average energy, which internalizes the factors of the  model. The last two terms specify entropies.
+
+Crucially, the BFE can be iteratively optimized for each individual variational distribution in turn. Optimization of the BFE is thus more manageable than direct optimization of the VFE. However, the BFE also entails an approximation of the full variational distribution by the Bethe approximation, which is not always exact.
```
Comment (bvdmitri) [inline docs/src/library/bethe-free-energy.md:64]: This line is a bit confusing,  the optimization of VFE and BFE entails an approximation, which is not always exact in both cases.

File: docs/src/library/bethe-free-energy.md [hunk 4]
```diff
@@ -1,5 +1,71 @@
...
+
+For iterative optimization of the BFE, the variational distributions must first be initialized. The `initmarginals` function initializes the variational distributions of the BFE.
```
Comment (bvdmitri) [inline docs/src/library/bethe-free-energy.md:66]: The `initmarginals` is not a function, but a keyword argument to the `inference` and `rxinference` function. You can use 
```
[`inference`](@ref) or [`rxinference`](@ref)
```
to generate links to their respective documentation.

**Review by bvdmitri — CHANGES_REQUESTED** [review]
Thanks @ThijsvdLaar! It's a good start, but there are a couple of comments I would like to be addressed before we can merge this. I noticed that some sentences are separated with blank lines and some are not, that makes a difference in the generated HTML file. 

For preview you can use `make docs` or `julia --project=docs docs/make.jl` (for the second command don't forget to initialize the `docs` environment). Either of these commands will generate HTML files in the `build` folder. Also you can check this small [guide](https://biaslab.github.io/RxInferDevGuides.jl/dev/guides/docs-contributing/#Use-[LiveServer.jl](https://github.com/tlienart/LiveServer.jl)).

 The PR also sais that it fixes the confusion between priors and initial marginals, but I didn't find that explanation in the text. For the rest look at my comments. Thanks!

@ismailsenoz can you also take a look (and perhaps add something from your side?)

**Comment by ThijsvdLaar** [comment]
Thanks for the feedback, I've made some changes to address your concerns.

@ismailsenoz might have more to say on these matters.

**Review by ismailsenoz — COMMENTED** [review]
I think it looks good. This level of detail is enough imo. It is a good summary @ThijsvdLaar. Maybe there is one thing that can be added. If the user chooses a constraint involving variables that are not argument to common factors it will be ignored by the inference engine due to Bethe assumption. There is always implicit Bethe constraints that the user should always be aware of.

**Comment by bvdmitri** [comment]
> If the user chooses a constraint involving variables that are not argument to common factors it will be ignored by the inference engine due to Bethe assumption. There is always implicit Bethe constraints that the user should always be aware of

I think this guide should not be a guide about `@constraints`, the guide about `@constraints` is a separate documentation page and we should keep in mind to mention it there https://github.com/biaslab/RxInfer.jl/issues/150

**Review by bvdmitri — COMMENTED** [review]
Thanks @ThijsvdLaar ! I like the text, the only thing which I think can be added is a list of relevant links to papers/materials for further reading. Feel free to add anything to the list what you find relevant. Then I can merge.

**Comment by ThijsvdLaar** [comment]
@bvdmitri I've added some of my go-to papers.

**Review by bvdmitri — APPROVED** [review]
Great, thanks!

**Comment by bvdmitri** [comment]
Lets see if tests are still passing......😁

**Pull request merged**